File: backend/src/main/java/com/yourcompany/onlineexam/controller/ExamSessionController.java
package com.yourcompany.onlineexam.controller;

import com.yourcompany.onlineexam.model.Part;
import com.yourcompany.onlineexam.service.ExamSessionService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Date;
import java.util.Map;

@RestController
@RequestMapping("/api/exam-session")
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:3000", "https://doanliennganh.vercel.app"})
public class ExamSessionController {
    private final ExamSessionService examSessionService;
    private static final Logger logger = LoggerFactory.getLogger(ExamSessionController.class);

    public ExamSessionController(ExamSessionService examSessionService) {
        this.examSessionService = examSessionService;
    }

    /**
     * Kiểm tra trạng thái bài thi (có thể làm bài hay không)
     */
    @GetMapping("/{partId}/status")
    public ResponseEntity<Map<String, Object>> getExamStatus(@PathVariable String partId) {
        try {
            Map<String, Object> status = examSessionService.getExamStatus(partId);
            return ResponseEntity.ok(status);
        } catch (Exception e) {
            logger.error("Lỗi khi kiểm tra trạng thái bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Bắt đầu làm bài thi
     */
    @PostMapping("/{partId}/start")
    public ResponseEntity<Map<String, Object>> startExam(
            @PathVariable String partId,
            @RequestBody Map<String, String> request) {
        try {
            String userEmail = request.get("userEmail");
            if (userEmail == null || userEmail.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of("error", "Thiếu thông tin userEmail"));
            }

            Map<String, Object> result = examSessionService.startExam(partId, userEmail);
            return ResponseEntity.ok(result);
        } catch (IllegalArgumentException e) {
            logger.warn("Lỗi validation khi bắt đầu bài thi {}: {}", partId, e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            logger.error("Lỗi khi bắt đầu bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Lấy danh sách bài thi có thể làm theo thời gian
     */
    @GetMapping("/available")
    public ResponseEntity<Map<String, Object>> getAvailableExams(
            @RequestParam String userEmail,
            @RequestParam(required = false) String courseId) {
        try {
            Map<String, Object> result = examSessionService.getAvailableExams(userEmail, courseId);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            logger.error("Lỗi khi lấy danh sách bài thi có thể làm: ", e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Kiểm tra thời gian còn lại của bài thi
     */
    @GetMapping("/{partId}/remaining-time")
    public ResponseEntity<Map<String, Object>> getRemainingTime(
            @PathVariable String partId,
            @RequestParam String userEmail) {
        try {
            Map<String, Object> result = examSessionService.getRemainingTime(partId, userEmail);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            logger.error("Lỗi khi kiểm tra thời gian còn lại bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }

    /**
     * Nộp bài thi
     */
    @PostMapping("/{partId}/submit")
    public ResponseEntity<Map<String, Object>> submitExam(
            @PathVariable String partId,
            @RequestBody Map<String, Object> request) {
        try {
            String userEmail = (String) request.get("userEmail");
            Map<String, Object> answers = (Map<String, Object>) request.get("answers");
            
            if (userEmail == null || userEmail.trim().isEmpty()) {
                return ResponseEntity.badRequest().body(Map.of("error", "Thiếu thông tin userEmail"));
            }

            Map<String, Object> result = examSessionService.submitExam(partId, userEmail, answers);
            return ResponseEntity.ok(result);
        } catch (IllegalArgumentException e) {
            logger.warn("Lỗi validation khi nộp bài thi {}: {}", partId, e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            logger.error("Lỗi khi nộp bài thi {}: ", partId, e);
            return ResponseEntity.status(500).body(Map.of("error", "Lỗi server"));
        }
    }
} 

File: backend/OnlineExam/Controllers/ExamSessionController.cs
using Microsoft.AspNetCore.Mvc;
using OnlineExam.Services;

namespace OnlineExam.Controllers;

[ApiController]
[Route("api/exam-session")]
public class ExamSessionController : ControllerBase
{
    private readonly ExamSessionService _service;
    private readonly ILogger<ExamSessionController> _logger;

    public ExamSessionController(ExamSessionService service, ILogger<ExamSessionController> logger)
    {
        _service = service;
        _logger = logger;
    }

    [HttpGet("{partId}/status")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetExamStatus(string partId)
    {
        try { return Ok(await _service.GetExamStatusAsync(partId)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi kiểm tra trạng thái bài thi {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpPost("{partId}/start")]
    public async Task<ActionResult<Dictionary<string, object?>>> StartExam(string partId, [FromBody] Dictionary<string, string> request)
    {
        try
        {
            if (!request.TryGetValue("userEmail", out var userEmail) || string.IsNullOrWhiteSpace(userEmail))
                return BadRequest(new Dictionary<string, object?> { ["error"] = "Thiếu thông tin userEmail" });
            return Ok(await _service.StartExamAsync(partId, userEmail));
        }
        catch (ArgumentException ex) { return BadRequest(new Dictionary<string, object?> { ["error"] = ex.Message }); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi bắt đầu bài thi {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("available")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetAvailableExams(
        [FromQuery] string userEmail, [FromQuery] string? courseId)
    {
        try { return Ok(await _service.GetAvailableExamsAsync(userEmail, courseId)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi lấy bài thi có thể làm"); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpGet("{partId}/remaining-time")]
    public async Task<ActionResult<Dictionary<string, object?>>> GetRemainingTime(string partId, [FromQuery] string userEmail)
    {
        try { return Ok(await _service.GetRemainingTimeAsync(partId, userEmail)); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi kiểm tra thời gian còn lại {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }

    [HttpPost("{partId}/submit")]
    public async Task<ActionResult<Dictionary<string, object?>>> SubmitExam(string partId, [FromBody] Dictionary<string, object?> request)
    {
        try
        {
            if (!request.TryGetValue("userEmail", out var userEmailObj) || userEmailObj?.ToString() is not { Length: > 0 } userEmail)
                return BadRequest(new Dictionary<string, object?> { ["error"] = "Thiếu thông tin userEmail" });

            Dictionary<string, object?>? answers = null;
            if (request.TryGetValue("answers", out var answersObj) && answersObj is Dictionary<string, object?> dict)
                answers = dict;
            else if (answersObj is System.Text.Json.JsonElement je && je.ValueKind == System.Text.Json.JsonValueKind.Object)
            {
                answers = je.EnumerateObject().ToDictionary(p => p.Name, p => (object?)p.Value.ToString());
            }

            return Ok(await _service.SubmitExamAsync(partId, userEmail, answers));
        }
        catch (ArgumentException ex) { return BadRequest(new Dictionary<string, object?> { ["error"] = ex.Message }); }
        catch (Exception ex) { _logger.LogError(ex, "Lỗi khi nộp bài thi {PartId}", partId); return StatusCode(500, new Dictionary<string, object?> { ["error"] = "Lỗi server" }); }
    }
}
